File: src/mahjong/models/player.py
from dataclasses import dataclass, field
from typing import List, Optional

from mahjong.constants.enums import Suit
from mahjong.models.meld import Meld
from mahjong.models.tile import Tile


@dataclass
class Player:
    player_id: str
    hand: List[Tile] = field(default_factory=list)
    melds: List[Meld] = field(default_factory=list)
    buried_cards: List[Tile] = field(default_factory=list)
    missing_suit: Optional[Suit] = None
    score: int = 100  # 初始分数100分（零和游戏：4人×100=400）
    is_hu: bool = False
    last_drawn_tile: Optional[Tile] = None  # 最后摸的牌（用于已胡玩家"摸什么打什么"）

File: frontend/src/types/player.ts
/**
 * Player state types
 *
 * Backend correspondence: src/mahjong/models/player.py::Player
 */

import { createTile, parseSuit } from './tile';
import { createMeld } from './meld';
import type { Tile, Suit } from './tile';
import type { Meld } from './meld';

/**
 * Player state
 *
 * Backend correspondence: @dataclass Player
 *
 * Note:
 * - hand field: Only visible for current player with complete hand, other players only show handCount
 * - missingSuit: Backend is missing_suit (needs conversion)
 * - isHu: Backend is is_hu (needs conversion)
 */
export interface Player {
  playerId: string;
  hand?: Tile[];              // Hand tiles (only visible for current player)
  handCount?: number;         // Hand tile count (visible for other players)
  melds: Meld[];              // Exposed tiles (visible to all)
  buriedCards: Tile[];        // Buried cards (visible to all)
  missingSuit: Suit | null;   // Missing suit
  score: number;              // Current score
  isHu: boolean;              // Whether already won (hu)
  lastDrawnTile?: Tile | null; // Last drawn tile (only visible for current player, used for self-drawn hu detection)
}

/**
 * Utility function: Create Player from backend data
 *
 * Backend returns two formats:
 * 1. Current player: includes hand (complete hand tiles)
 * 2. Other players: includes hand_count (hand tile count)
 */
export function createPlayer(data: any): Player {
  return {
    playerId: data.player_id,
    hand: data.hand ? data.hand.map(createTile) : undefined,
    handCount: data.hand_count,
    melds: (data.melds || []).map(createMeld),
    buriedCards: (data.buried_cards || []).map(createTile),
    missingSuit: data.missing_suit ? parseSuit(data.missing_suit) : null,
    score: data.score,
    isHu: data.is_hu,
    lastDrawnTile: data.last_drawn_tile ? createTile(data.last_drawn_tile) : null
  };
}
